src/ibmca_ec.c

```diff
@@ -0,0 +1,918 @@
+ #ifdef OLDER_OPENSSL
+        return 0;
+ #else
+        EC_KEY_METHOD_get_compute_key(ossl_ec, &compute_key_sw);
+        if (compute_key_sw == NULL) {
+            IBMCAerr(IBMCA_F_IBMCA_ECDH_COMPUTE_KEY,
+                     IBMCA_R_EC_INTERNAL_ERROR);
+            return 0;
+        }
+
+        return compute_key_sw(pout, poutlen, pub_key, ecdh);
+ #endif
+    }
+
+    /* Get (X,Y) from EC_POINT */
+    bn_x = BN_new();
+    bn_y = BN_new();
+    if (!EC_POINT_get_affine_coordinates_GFp(group, pub_key, bn_x, bn_y, NULL)) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDH_COMPUTE_KEY, IBMCA_R_EC_INTERNAL_ERROR);
+        goto end;
+    }
+
+    /* Format (X) as char array, with leading zeros if necessary */
+    n = privlen - BN_num_bytes(bn_x);
+    for (i = 0; i < n; i++)
+        X[i] = 0x00;
+    BN_bn2bin(bn_x, &(X[n]));
+
+    /* Format (Y) as char array, with leading zeros if necessary */
+    n = privlen - BN_num_bytes(bn_y);
+    for (i = 0; i < n; i++)
+        Y[i] = 0x00;
+    BN_bn2bin(bn_y, &(Y[n]));
+
+    /* Initialize public ICA_EC_KEY with (X,Y) */
+    rc = p_ica_ec_key_init(X, Y, NULL, ica_pubkey);
+    if (rc != 0) {
+        IBMCAerr(IBMCA_F_ICA_EC_KEY_INIT, rc);
+        goto end;
+    }
+
+    /* Create ICA_EC_KEY object for private key */
+    ica_privkey = p_ica_ec_key_new(nid, &privlen);
+    if (!ica_privkey) {
+        IBMCAerr(IBMCA_F_ICA_EC_KEY_NEW, IBMCA_R_EC_INTERNAL_ERROR);
+        goto end;
+    }
+
+    /* Get private (D) value from EC_KEY */
+    bn_d = (BIGNUM*)EC_KEY_get0_private_key((EC_KEY*)ecdh);
+    if (bn_d == NULL) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDH_COMPUTE_KEY, IBMCA_R_EC_INTERNAL_ERROR);
+        goto end;
+    }
+
+    /* Format (D) as char array, with leading zeros if necessary */
+    n = privlen - BN_num_bytes(bn_d);
+    for (i = 0; i < n; i++)
+        D[i] = 0x00;
+    BN_bn2bin(bn_d, &(D[n]));
+
+    /* Initialize private ICA_EC_KEY with (D) */
+    rc = p_ica_ec_key_init(NULL, NULL, D, ica_privkey);
+    if (rc != 0) {
+        IBMCAerr(IBMCA_F_ICA_EC_KEY_INIT, rc);
+        goto end;
+    }
+
+    /* Allocate memory for shared secret z, will be freed by caller */
+    if ((z_buf = OPENSSL_malloc(privlen)) == NULL) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDH_COMPUTE_KEY, IBMCA_R_EC_INTERNAL_ERROR);
+        goto end;
+    }
+
+    /* Calculate shared secret z */
+    rc = p_ica_ecdh_derive_secret(ibmca_handle, ica_privkey, ica_pubkey, z_buf,
+                                  privlen);
+    if (rc != 0) {
+        /* Possibly no suitable adapter. */
+        OPENSSL_free(z_buf);
+
+ #ifdef OLDER_OPENSSL
+        goto end;
+ #else
+        EC_KEY_METHOD_get_compute_key(ossl_ec, &compute_key_sw);
+        if (compute_key_sw == NULL) {
+            IBMCAerr(IBMCA_F_ICA_ECDH_DERIVE_SECRET, rc);
+            goto end;
+        }
+
+        ret = compute_key_sw(pout, poutlen, pub_key, ecdh);
+        goto end;
+ #endif
+    }
+
+    *pout = z_buf;
+    *poutlen = privlen;
+
```

Comment:
when the branch with software fallback above is used, pout and poutlen have their values already and are now overwritten with wrong values. Fix it !

Reply:
forget it, my thoughts where wrong